### ISSUE TITLE
Indicate what the voltmeter probes are touching

### ISSUE THREAD

**Issue by samreid** [phetsims/capacitor-lab-basics#223]
For #218 it was requested we add an event to the data stream that indicates what the voltmeter probes are contacting

> - When I do measurements using the voltmeter, I get the position (x,y,z) of the positive and negative probes. Is it possible to map those positions onto the sim elements? I would for example be interested in knowing, that the student is trying to measure the voltage on the capacitor, or the battery, the light bulb...

I replied:

> Good idea, I'll check into the code to see how difficult this will be to add, and will discuss with Kathy to see if we have time to implement it before the study.

**Comment by samreid** [comment]
It seems we should add something like this in Voltmeter.js:

```js
    var negativeProbeContactProperty = new Property( null, {
      tandem: tandem.createTandem( 'negativeProbeContactProperty' ),
      phetioType: PropertyIO( NullableIO( StringIO ) )
    } );
```

@jonathanolson it looks like the logic for hit testing is implemented in Voltmeter.computeValue with a fallback to circuit.getVoltageBetween.  I'm hesitant to add checks for each intersection at the beginning of Voltmeter.computeValue because it may have performance implications (correct me if that's wrong).  Do we need to return multiple values for intersection?  I.e. what if the negative probe intersects the battery anode as well as the positive probe?  The physics is based on the short circuit--so should we put assignments by each return value? But then it would be pretty scattered around (I guess it already is somewhat scattered).

Also, it may be nice to use `CircuitLocation` if is is a complete set of possible intersections of the voltmeter probe (may need to add "NEGATIVE_PROBE" and "POSITIVE_PROBE"), so the possible values and documented and standardized.

To summarize, I see a few ways we could do this, but not the best way to implement it elegantly.  @jonathanolson can you suggest how to proceed?

**Comment by samreid** [comment]
One way to proceed would be to have a `getConnection` function for each probe, which returns a CircuitConnection object, which has an identifier (name), voltage relative to ground, and a unique ground identifier.  Then to compare voltages between two Connection instances, you would have this pseudocode:

if (c1.ground===c2.ground){
return c2.value - c1.value;
}
else return null;

This would generalize most of the logic in Voltmeter.computeValue, and we could use the strings for the data stream (I think we'd still need a special case for probe/probe intersection).  On the other hand, this would be a non-trivial rewrite of about 100 lines of critical code.

**Comment by samreid** [comment]
Also assigning @kathy-phet so she's aware of some of the possibilities, and can comment on whether we can/should tackle it before the study.

**Comment by jonathanolson** [comment]
> I'm hesitant to add checks for each intersection at the beginning of Voltmeter.computeValue because it may have performance implications (correct me if that's wrong).

computeValue needs to be called a lot (once per frame during changes), so presumably if we are handling the connection status once per frame (computing it, and using it for computing the voltage value) then we should be fine.

> Do we need to return multiple values for intersection? I.e. what if the negative probe intersects the battery anode as well as the positive probe?

I believe there wasn't a need for this in the code, since if the positive probe touches the negative probe then it can ignore where it may be touching the circuit (and everything else is spaced hopefully far enough apart?). Depends how researchers/users would want the data stream to be.

> Also, it may be nice to use CircuitLocation if is is a complete set of possible intersections of the voltmeter probe (may need to add "NEGATIVE_PROBE" and "POSITIVE_PROBE"), so the possible values and documented and standardized.

That may be an option, but presumably you would want to differentiate between "over top of the battery" and "over the wire that connects to the top of the battery" also? I wish I had a better answer, but I'm slightly foggier on details since I wasn't handling the initial implementation. Could collaborate on ways to handle this.

**Comment by samreid** [comment]
Should we change the return value from {number} to {negativeProbeContact:{string},voltage:{number},positiveProbeContact:{string}}, then we can use the contact points that are pertinent in the physics calculation?

**Comment by samreid** [comment]
@jonathanolson and I discussed it and it seems we can retrofit the existing code to include a new "ConnectionPoint" enum rather than rewriting it from scratch.  That may be less risky.

I think it could take 2-4 hours for me to get this to a first draft where I think it is probably working.  @jonathanolson and I can work on this together to reduce risk even further.

**Comment by samreid** [comment]
@jonathanolson and I got a start on this, and @jonathanolson volunteered to take the lead on this.

**Comment by jonathanolson** [comment]
@arouinfar and @kathy-phet, how do we want to handle the case where a probe tip can technically overlap over two different wires:
![clb-connection](https://user-images.githubusercontent.com/480878/35234277-ca9f18be-ff5d-11e7-84bc-8de7ee9886e8.PNG)

It would be simpler to say that it overlaps just one thing, and report that. However if that is the case, and I rewrite logic to be based on the "one thing the probe tip overlaps", then this example wouldn't count as being over both.

Additionally, it seems that the probe tips would complete a circuit if you get them to overlap like in that case? That is not handled in the code, and hopefully we don't need to handle it. You can also overlap probe tips so that it would connect the capacitor plates:
![clb-both-plates](https://user-images.githubusercontent.com/480878/35234523-b50ab818-ff5e-11e7-92a7-e2d4458d504a.PNG)

**Comment by jonathanolson** [comment]
Unassigned myself, as I'll need to know whether the slight sim behavior change (where the probe COULD be overlapping two things, and both are checked now, but it would only think it would be over one in the future) is fine, and/or whether we can report that for phet-io.

**Comment by samreid** [comment]
The dev test completed successfully in https://github.com/phetsims/QA/issues/85, and hence this is the only issue I'm aware of that would prevent next dev or rc test.  Elevating priority to get feedback from @kathy-phet or @arouinfar before @jonathanolson can proceed.

**Comment by arouinfar** [comment]
@jonathanolson @samreid I think this is entirely @kathy-phet's call, and not mine.  I'm not that familiar with the client's needs, but  I'll go ahead and throw in my two cents.

> how do we want to handle the case where a probe tip can technically overlap over two different wires:
![image](https://user-images.githubusercontent.com/8419308/35302783-fc89c168-004c-11e8-9677-cf4a35de7b16.png)
It would be simpler to say that it overlaps just one thing, and report that.

@jonathanolson in my mind, the probe tip is only overlapping one thing -- the solder joint. If you report that the probe tip is overlapping both "wire A" and "wire B", then I think it's fairly obvious that the probe must be on an intersection of sorts. 

> Additionally, it seems that the probe tips would complete a circuit if you get them to overlap like in that case? That is not handled in the code, and hopefully we don't need to handle it

I do not think we want to handle cases like that.

**Comment by kathy-phet** [comment]
@jonathanolson - It might be easiest to try to zoom to discuss, as I'm not entirely sure I understand your questions.

**Comment by jonathanolson** [comment]
Zoom would probably be best to discuss, as there are a few semi-similar but different issues that are somewhat hard to describe, and would help way more to be able to screen-share.

What would be a good time (if possible)?

**Comment by jonathanolson** [comment]
Implemented the "probes are only touching one thing" full strategy (including for voltage computation), using priority so that if a probe is technically over multiple things, it will use the "more obvious" component (e.g. the capacitor top plate has precedence over its adjacent wire).

@samreid, can you check briefly over my changes, and make sure the phet-io events and behavior looks fine?

I'd also HIGHLY recommend QA testing for regressions, checking all different probe location combinations in all situations. The commits reduced things by ~180 lines of code, so I may have missed a combination.

**Comment by samreid** [comment]
I noticed that `this.positiveProbeTargetProperty` and `this.negativeProbeTargetProperty` are marked as `@public` but they are not used outside of Voltmeter.js.  Consider changing them to locally scoped vars and passing them through as arguments to `computeValue`

**Comment by samreid** [comment]
This code appears twice:

```js
      // If the probes are touching the same location, there should be no voltage change
      if ( positiveCircuitLocation === negativeCircuitLocation ) {
        return 0;
      }
```

**Comment by samreid** [comment]
The rest of the cursory code review looks great, and the data stream looks great, nice work!  Leaving assigned to @jonathanolson about the preceding two comments.

**Comment by phet-steele** [comment]
@jonathanolson @samreid this is looking great in 1.6.4 phet-io after testing it for 3 hours, will leave open for phet brand testing, or in case a bug finally reveals itself.

**Comment by phet-steele** [comment]
Probes are working well. @jonathanolson please see https://github.com/phetsims/capacitor-lab-basics/issues/223#issuecomment-360596757 and https://github.com/phetsims/capacitor-lab-basics/issues/223#issuecomment-360597402

**Comment by jonathanolson** [comment]
I addressed both comments (it is technically not a duplicate check, and both are there for a purpose, but documented now). Does this look good @samreid?

**Comment by samreid** [comment]
It looks great, thanks!  Closing.
